Introduce support for QTEE supplicant daemon and listeners

harshaldev27 opened the pull request:
This Pull Request introduces support for the QTEE supplicant daemon, including:

1. Time Listener.
2. TA Autoload Listener.
3.
The QTEE supplicant daemon continuously waits for requests from QTEE for REE-provided services (e.g., system time). Upon receiving a request, it invokes the corresponding listener functionality and returns a response to QTEE.

Each listener service is provided by a dynamically linked shared object library, which is loaded and registered at runtime by the QTEE supplicant from a predefined list of expected listener libraries available on the device. Listener services can also provide their own registration callback if they prefer not to use the default listener registration provided by the supplicant.

This Pull Request has been tested on the Qualcomm SM8650-MTP platform using the smcinvoke_client and gp_test_client.

quicbasantk commented on qtee_supplicant/include/heap.h:
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
```
can we remove this file?
I feel use stdlib  API is cleaner compared to creating own wrapper?

quicbasantk commented on qtee_supplicant/include/heap_port.h:
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: BSD-3-Clause-Clear
```
can we remove this header?

quicbasantk commented on qtee_supplicant/include/listener_mngr.h:
```diff
@@ -0,0 +1,52 @@
+	void *lib_handle; /**< LibHandle for the listener. */
+	char *svc_init; /**< Listener service (opt) init callback. */
+	char *smci_dispatch; /**< Dispatch function of the listener. */
+	Object cbo; /**< Store cbo of listener service. */
```
do we really need to keep track of cbo and smo in listener_svc if is tracked through Clistener context?

quicbasantk commented on qtee_supplicant/include/listener_mngr.h:
```diff
@@ -0,0 +1,52 @@
+struct listener_svc {
+	char *service_name; /**< Name of the listener service. */
+	int id; /**< Id of the listener service. */
+	int is_registered; /**< Listener registration status. */
```
this field looks redundant, valid cbo init sequence can be changed to signify listener is registered?

quicbasantk commented on qtee_supplicant/include/listener_mngr.h:
```diff
@@ -0,0 +1,52 @@
+	char *file_name; /**< File name of the listener service. */
+	void *lib_handle; /**< LibHandle for the listener. */
+	char *svc_init; /**< Listener service (opt) init callback. */
+	char *smci_dispatch; /**< Dispatch function of the listener. */
```
can we rename to dispatcher?

harshaldev27 replied:
Ack.

quicbasantk commented on qtee_supplicant/src/qtee_supplicant.c:
```diff
@@ -0,0 +1,34 @@
+
+	MSGD("qtee_supplicant: process entry PPID = %d\n", getppid());
+
+	signal(SIGPIPE, SIG_IGN);
```
do we have any significance of this signal?

quicbasantk commented on qtee_supplicant/src/listener_mngr.c:
```diff
@@ -0,0 +1,212 @@
+
+#define NUM_LISTENERS 1
+static struct listener_svc listeners[] = {
+	{
```
Please ensure each listener is feature controlled.
We should not assume any listener to default.

quicbasantk commented on qtee_supplicant/src/listener_mngr.c:
```diff
@@ -0,0 +1,212 @@
+static struct listener_svc listeners[] = {
+	{
+		.service_name = "time services",
+		.id = TIME_SERVICE,
```
I am thinking, can we move this struct static config to each domain listener?
In this file we will maintain one struct list which has only list of ids based on feature flag and introudce one new function in each listener get_config() to populate this array dyanamically?

quicbasantk commented on qtee_supplicant/src/listener_mngr.c:
```diff
@@ -0,0 +1,212 @@
+		.file_name = "libdrmtime.so.1",
+		.lib_handle = NULL,
+		.svc_init = NULL,
+		.smci_dispatch = "smci_dispatch",
```
if dispatcher name is same in all listeners, do we really want to track here?

quicbasantk commented on qtee_supplicant/include/qlist.h:
```diff
@@ -0,0 +1,209 @@
+    constant-time insertion or removal of items or of entire queues.
+    Because it is circular it permits constant-time operations at both the
+    tail and the head of the queue.  Circularity also streamlines some
+    operations by eliminating conditional branches.
```
can we think of using any std lib for qlist?

quicbasantk commented on listeners/libdrmtime/atime.c:
```diff
@@ -0,0 +1,307 @@
+	return 0;
+}
+
+int smci_dispatch(void *dma_buf, size_t dma_buf_len)
```
can we move all functions except dispatcher to another file atime_operation.c and call from dispatcher?
This will help to streamline in case we have to support legacy qseecom as well?

quicbasantk reviewed (commented):
This looks good to me, just check do we need to copy both skel and interface header file?
Or no harm generating and copying both?

harshaldev27 commented:
Changes in patchset 2:
1. Introduced flags for conditional compilation of individual listeners.
2. Re-named smci_dispatch_func to dispatch_func.
3. Removed heap_port.h and heap.h
4. Removed 'smo' member from listener_svc.
5. Removed SIGPIPE signal ignore from qtee_supplicant.c
6. Fixed memory leak in tmp_buf.

Changes not introduced in patchset 2:
1. Did not introduce get_info() function yet, needs more discussion since it might not be scalable.
2. Did not yet combine init() and register() into one function along with movement of CListenerCBO and default registration into separate library. This appears to be an overkill design for solving a not so immediate problem. Needs further discussion.
